[BOLT][AArch64] Make gs-pacret-autiasp.s deterministic

paschalis-mpeis opened the pull request:
In gs-pacret-autiasp.s, the undefined call `bl g` causes inconsistent
basic block splitting: in some platforms BOLT emits two blocks, on some
others one.

Defining a dummy `g` symbol forces a single basic block everywhere.

paschalis-mpeis committed:
[BOLT][AArch64] Skip gadget pacret test on RHEL8

paschalis-mpeis committed:
fix: py code formatter

kbeyls commented:
If I understand correctly, this test has assembly as input, where we expect the exact same binary to be produced on all platforms. Also, the analyzer is expected to produce the same results, given a particular input binary, on all platforms.

For that reason, it seems to me that disabling this test on RHEL8 is the wrong fix. Would you be able to show a the input to FileCheck, so that a more meaningful diff can be produced, and investigate what the root cause is of why this test produces a different result on RHEL8?

paschalis-mpeis commented:
Hey Kristof,

Thanks for your review, I agree.
I'm marking this PR as draft and I'll ask internally for more details on the root cause.
I'll follow up once there's more clarity.

paschalis-mpeis commented:
Hey @atrosinenko,

Thanks a lot for your comment. Reading this back to see if I got it right.
**(1)** you are saying the following should (probably) work on RHEL8:

```diff
// CHECK-LABEL: GS-PAUTH: non-protected ret found in function f1, basic block {{[0-9a-zA-Z.]+}}, at address
// CHECK-NEXT:    The instruction is     {{[0-9a-f]+}}:       ret
// CHECK-NEXT:    The 1 instructions that write to the affected registers after any authentication are:
// CHECK-NEXT:    1. {{[0-9a-f]+}}: ldp     x29, x30, [sp], #0x10
// CHECK-NEXT:  This happens in the following basic block:
// CHECK-NEXT: {{[0-9a-f]+}}:   paciasp
+ // CHECK-NEXT: {{[0-9a-f]+}}:   stp     x29, x30, [sp, #-16]!
+ // CHECK-NEXT: {{[0-9a-f]+}}:   mov     x29, sp
+ // CHECK-NEXT: {{[0-9a-f]+}}:   bl      g
// CHECK-NEXT: {{[0-9a-f]+}}:   add     x0, x0, #0x3
// CHECK-NEXT: {{[0-9a-f]+}}:   ldp     x29, x30, [sp], #0x10
// CHECK-NEXT: {{[0-9a-f]+}}:   ret
        ret
```


**(2)** you're asking why on other platforms (non-RHEL8) a `bl` call splits the basic block?

So we should expect BOLT IR to produce a single block for `f1`. Given we achieve this consistently, we could add the above extra 3 lines to the test?


I can confirm that on my platform (a non-RHEL8 linux) I get two basic blocks in `f1`, where the `g` symbol is undefined. If replace the call target with something that exists (ie `f_intermediate_overwrite1`), I get a single block.

kbeyls commented:
Thanks for the continued analysis @atrosinenko and @paschalis-mpeis !
This rings a bell to me now.
When I landed the initial pac-ret scanner, the bots complained because basic blocks were split differently compared to what I observed at the time locally on my slightly older version of BOLT. See the discussion at https://github.com/llvm/llvm-project/pull/128576

paschalis-mpeis committed:
[BOLT][AArch64] Make gs-pacret-autiasp.s deterministic

In gs-pacret-autiasp.s, the undefined call `bl g` causes inconsistent
basic block splitting: in some platforms BOLT emits two blocks, on some
others one.

Defining a dummy `g` symbol forces a single basic block everywhere.

kbeyls reviewed (approved):
Thanks, this looks like an acceptable work-around!

paschalis-mpeis commented:
Hey @kbeyls and @atrosinenko,

That was really helpful, thank you! 🙏 

 I've added the `g` symbol and updated each test case affected to make the test deterministic.

The root cause of the unresolved symbol [behaviour](https://github.com/llvm/llvm-project/pull/128576#issuecomment-2682489666) is still unclear. Perhaps BOLT treated that undefined call as non-external and assumes it cannot return/recover, causing the block to split?

We should keep this in mind and at some point investigate it further.

atrosinenko reviewed (approved):
LGTM as long as this still passes on other builders, thank you!

I observed some issues like this in the past, too. IIRC there were problems with calls performed via PLT. Furthermore, IIRC it may behave differently depending on whether `--emit-relocs` is passed to the linker, as suggested in the documentation of BOLT.

paschalis-mpeis commented:
Thanks Kristof and Anatoly for the help and reviews!

cc @maksfb so he's aware of this non-determinism (here's previous [related discussion](https://github.com/llvm/llvm-project/pull/128576#issuecomment-2680524529) with @kbeyls).
